packages/cli/src/databases/repositories/execution.repository.ts

```diff
@@ -399,21 +413,90 @@ export class ExecutionRepository extends Repository<ExecutionEntity> {
 
 		if (!executions.length) {
 			if (deleteConditions.ids) {
-				LoggerProxy.error('Failed to delete an execution due to insufficient permissions', {
+				Logger.error('Failed to delete an execution due to insufficient permissions', {
 					executionIds: deleteConditions.ids,
 				});
 			}
 			return;
 		}
 
 		const executionIds = executions.map(({ id }) => id);
-		const binaryDataManager = BinaryDataManager.getInstance();
-		await binaryDataManager.deleteBinaryDataByExecutionIds(executionIds);
-
 		do {
 			// Delete in batches to avoid "SQLITE_ERROR: Expression tree is too large (maximum depth 1000)" error
-			const batch = executionIds.splice(0, 500);
-			await this.delete(batch);
+			const batch = executionIds.splice(0, this.deletionBatchSize);
+			await this.softDelete(batch);
 		} while (executionIds.length > 0);
 	}
+
+	async pruneBySoftDeleting() {
+		Logger.verbose('Pruning (soft-deleting) execution data from database');
+
+		const maxAge = config.getEnv('executions.pruneDataMaxAge'); // in h
+		const maxCount = config.getEnv('executions.pruneDataMaxCount');
+
+		// Find ids of all executions that were stopped longer that pruneDataMaxAge ago
+		const date = new Date();
+		date.setHours(date.getHours() - maxAge);
+
+		const toPrune: Array<FindOptionsWhere<ExecutionEntity>> = [
+			// date reformatting needed - see https://github.com/typeorm/typeorm/issues/2286
+			{ stoppedAt: LessThanOrEqual(DateUtils.mixedDateToUtcDatetimeString(date)) },
+		];
+
+		if (maxCount > 0) {
+			const executions = await this.find({
+				select: ['id'],
+				skip: maxCount,
+				take: 1,
+				order: { id: 'DESC' },
+			});
+
+			if (executions[0]) {
+				toPrune.push({ id: LessThanOrEqual(executions[0].id) });
+			}
+		}
+
+		const [timeBasedWhere, countBasedWhere] = toPrune;
+
+		await this.createQueryBuilder()
+			.update(ExecutionEntity)
+			.set({ deletedAt: new Date() })
+			.where(
+				new Brackets((qb) =>
+					countBasedWhere
+						? qb.where(timeBasedWhere).orWhere(countBasedWhere)
+						: qb.where(timeBasedWhere),
+				),
+			)
+			.execute();
+	}
+
+	/**
+	 * Permanently delete all soft-deleted executions and their binary data, in batches.
+	 */
+	private async hardDelete() {
+		// Find ids of all executions that were deleted over an hour ago
+		const date = new Date();
+		date.setHours(date.getHours() - 1);
+
+		const executionIds = (
+			await this.find({
+				select: ['id'],
+				where: {
+					deletedAt: LessThanOrEqual(DateUtils.mixedDateToUtcDatetimeString(date)),
+				},
+				take: this.deletionBatchSize,
+			})
+		).map(({ id }) => id);
+
+		const binaryDataManager = BinaryDataManager.getInstance();
+		await binaryDataManager.deleteBinaryDataByExecutionIds(executionIds);
+
+		// Actually delete these executions
+		await this.delete({ id: In(executionIds) });
+
+		if (executionIds.length === this.deletionBatchSize) {
+			setTimeout(async () => this.hardDelete(), 1000);
```

Comment:
One thing that worries me with this strategy is that n8n might start piling up those if the instance runs too many executions (say > 100/s - this would pile up the pruning as well).

Maybe we should consider stopping the 15 minutes interval from firing if this one is happening. so basically:

```
if (executionIds.length === this.deletionBatchSize) {
    // Stop the 15 minutes timer (cancel the interval)
    // Schedule this function to run every 1 second, like the above
} else {
    // If 15 minutes interval is up, keep it.
    // If not up, set it
}
```

This would prevent possible collisions meaning n8n would try to delete the same execution twice, causing possible concurrency errors, since we fetch 100 IDs and work on them 1 by 1.

Reply:
Thank you, please [0e075c2](https://github.com/n8n-io/n8n/pull/7092/commits/0e075c257d26f5f34fb722fdd2f5120a96f09023)